### PR TITLE
feat: #1174 add tests for get my fam access api

### PR DESCRIPTION
**Pull request by ianliuwk1019**
Tests added for get_admin_user_access.
- Test with user that has no role
- Test with user that has only FAM-ADMIN role
- Test with user that has only/multiple app admin roles
- Test with user that has only/multiple delegated admin access (ie; FOM_SUBMITTER)
- Test with user that has multiple access privileges.

### CODE REVIEW

File: server/admin_management/tests/conftest.py [hunk 1]
```diff
@@ -130,6 +140,126 @@ def test_rsa_key_missing():
     return new_key.exportKey("PEM")
 
 
+@pytest.fixture(scope="function")
+def setup_new_user(user_repo: UserRepository, db_pg_session: Session):
```
Comment (MCatherine1994) [inline server/admin_management/tests/conftest.py:144]: I wonder if we could add the return type for this method? as well as the setup_new_fom_delegated_admin?

File: server/admin_management/tests/conftest.py [hunk 2]
```diff
@@ -130,6 +140,126 @@ def test_rsa_key_missing():
...
+
+
+@pytest.fixture(scope="function")
+def setup_new_fom_delegated_admin(
+    access_control_privilege_repo: AccessControlPrivilegeRepository,
+    role_service: RoleService
+):
+    """
+    Conveniently setup new "FOM" DELEGATED_ADMIN user for testing.
+    The fixture returns a function to be called with new FOM (evn) delegated
```
Comment (MCatherine1994) [inline server/admin_management/tests/conftest.py:210]: just a typo "evn"

File: server/admin_management/tests/routers/test_router_admin_user_access.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    fom_test_roles_grants = fom_test_grant["roles"]
+    # was setup for ROLE_TYPE_ABSTRACT (FOM_SUBMITTER), so should only have 1
+    assert len(fom_dev_roles_grants) == 1
+    fom_test_role_grants_submitter = fom_test_roles_grants[0]
+    assert fom_test_role_grants_submitter["id"] == TEST_FOM_TEST_SUBMITTER_ROLE_ID
+    assert fom_test_role_grants_submitter["name"] == TEST_APP_ROLE_NAME_FOM_SUBMITTER
+    assert fom_test_role_grants_submitter["type_code"] == RoleType.ROLE_TYPE_ABSTRACT.value
+    # should contains exact 2 forest_clients the new_user can administer
+    assert set(fom_test_role_grants_submitter["forest_clients"]) == \
+        set([TEST_FOREST_CLIENT_NUMBER, TEST_FOREST_CLIENT_NUMBER_TWO])
```
Comment (MCatherine1994) [inline server/admin_management/tests/routers/test_router_admin_user_access.py:274]: maybe we can use "dga_submitter_forest_clients" here?

File: server/admin_management/tests/services/test_admin_user_access_service.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
+import logging
+
+import pytest
```
Comment (MCatherine1994) [inline server/admin_management/tests/services/test_admin_user_access_service.py:3]: this import is not in use

File: server/admin_management/tests/routers/test_router_admin_user_access.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+                             TEST_FOM_TEST_SUBMITTER_ROLE_ID,
+                             TEST_FOREST_CLIENT_NUMBER,
+                             TEST_FOREST_CLIENT_NUMBER_TWO,
+                             TEST_NEW_BCEID_USER, TEST_NEW_IDIR_USER)
+
+LOGGER = logging.getLogger(__name__)
+test_end_point = f"{apiPrefix}/admin-user-accesses"
+
+# Below tests are based on endpoint json sample output at confluence design:
+# https://apps.nrs.gov.bc.ca/int/confluence/display/FSAST1/Delegated+Access+Administration+Design
```
Comment (MCatherine1994) [inline server/admin_management/tests/routers/test_router_admin_user_access.py:28]: Hi Ian, I wonder if you want to update this confluence page to make the json sample up to date? For example, in the below tests, for app admin, roles will have forest client number None, `assert fom_dev_roles_grants_submitter["forest_clients"] is None`, but it's not shown in the json sample. Think the json example in your last pr is the latest https://github.com/bcgov/nr-forests-access-management/pull/1170
Screen Shot 2024-02-12 at 4 06 48 PM
Screen Shot 2024-02-12 at 4 06 58 PM

File: server/admin_management/tests/repositories/test_access_control_privilege_repository.py [hunk 6]
```diff
@@ -147,3 +146,53 @@ def test_get_acp_by_application_id(
         TEST_NOT_EXIST_APPLICATION_ID
     )
     assert len(found_acp) == 0
+
+
+def test_get_user_delegated_admin_grants(
+        access_control_privilege_repo: AccessControlPrivilegeRepository,
+        user_repo: UserRepository
```
Comment (MCatherine1994) [inline server/admin_management/tests/repositories/test_access_control_privilege_repository.py:153]: maybe we format this file?

File: server/admin_management/tests/routers/test_router_admin_user_access.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    assert response.status_code == HTTPStatus.OK
+    result = response.json().get("access")
+    assert len(result) == 1
+    access = result[0]
+    assert access.get("auth_key") == AdminRoleAuthGroup.FAM_ADMIN  # FAM_ADMIN grants
+    grants = access.get("grants")
+    fam_app_grant = list(filter(lambda x: x["application"]["id"] == TEST_APPLICATION_ID_FAM, grants))[0]
+    assert fam_app_grant["application"] is not None
+    assert fam_app_grant["application"]["id"] == TEST_APPLICATION_ID_FAM
+    assert fam_app_grant["application"]["name"] == TEST_APPLICATION_NAME_FAM  # verify grante app is "FAM", not "FAM_DEV/FAM_TEST"
```
Comment (MCatherine1994) [inline server/admin_management/tests/routers/test_router_admin_user_access.py:91]: maybe it's a typo 'grante'

File: server/admin_management/tests/routers/test_router_admin_user_access.py (hunk 5 above)
Comment (ianliuwk1019) [reply server/admin_management/tests/routers/test_router_admin_user_access.py:28]: Good suggestion. I updated the confluence.

However, on the side note, it actually has a layer of translation between python (FastAPI) and Swagger before outputting the swagger return json we see on api's doc (on swagger). I noticed using Python's "Optional" and "Union" types can have difference at the swagger's translation (for example, difference between: None becomes "null" or None becomes => the key/alue being removed). If we do need to make it: None becomes => the key/alue being removed, we can have another ticket to adjust (but not clear what to adjust, the python type, or is there existing swagger conversion configuration for this?).

Comment (MCatherine1994) [reply server/admin_management/tests/routers/test_router_admin_user_access.py:28]: I would prefer to make it optional, so if "None" then remove key/value pair

**Review by MCatherine1994 — APPROVED** [review]

File: server/admin_management/tests/routers/test_router_admin_user_access.py (hunk 5 above)
Comment (ianliuwk1019) [reply server/admin_management/tests/routers/test_router_admin_user_access.py:28]: Thanks Catherine.
I also prefer "Optional" than "Union" in this case.
However, it is the "Optional" that Swagger is producing the "null", not removing the key/value pair.
This is the issue between FastAPI(python) and Swagger, not really in our code.
I will create a "nice-to-have tech debt" ticket so we can research/investigate if there is a way to tweak FastAPI for this at current or higher version.

**Pull request merged**